README.md
[![Latest Stable Version](https://poser.pugx.org/tomgrohl/laravel4-php71-encrypter/v/stable)](https://packagist.org/packages/tomgrohl/laravel4-php71-encrypter)
[![Latest Unstable Version](https://poser.pugx.org/tomgrohl/laravel4-php71-encrypter/v/unstable)](https://packagist.org/packages/tomgrohl/laravel4-php71-encrypter)
[![License](https://poser.pugx.org/tomgrohl/laravel4-php71-encrypter/license)](https://packagist.org/packages/tomgrohl/laravel4-php71-encrypter)

# Laravel 4.2 Encrypter for PHP 7.1+

A port of the Laravel 5.4 Encrypter for 4.2 with PHP 7.1 support.

Due to mcrypt being deprecated in 7.1 you can't run Laravel 4.2 under PHP 7.1 without a lot of deprecate warnings.

This package adds a PHP 7.1+ Encrypter for Laravel 4.2

> **Note:** Backward compatibility has been added in version 1.1.0 for PHP 5.6+ to make upgrading easier

## Installation

You can install it using composer:

`composer require tomgrohl/laravel4-php71-encrypter`


## Configuration

### 1 .Add service provider

Add the following to your `providers` in the `app` config

```
\Tomgrohl\Laravel\Encryption\EncryptionServiceProvider::class
```

### 2. Check Encryption Key settings in app config

Cipher must either be 'AES-128-CBC' with a key length of 16

OR     
       
Cipher must either be 'AES-256-CBC' with a key length of 32


### 3.  You're good to go!


# Credits

- Peter Fritchley - Fellow co-worker that wrote this originally for a project

# License

MIT

CHANGELOG.md
1.1.3
    - Bug fixes
    
1.1.2
    - Bug fixes
    
1.1.1
    - Bug fixes
    
1.1.0
    - Made backward compatible with PHP 5.6+ to make upgrading easier

1.0.0
    - Initial Release


src/Tomgrohl/Laravel/Encryption/Tests/TestCase.php
<?php

namespace Tomgrohl\Laravel\Encryption\Tests;

use PHPUnit_Framework_TestCase;

/**
 * Class TestCase
 *
 * @package Tomgrohl\Laravel\Encryption\Tests
 */
abstract class TestCase extends PHPUnit_Framework_TestCase
{

}


src/Tomgrohl/Laravel/Encryption/Tests/EncrypterTest.php
<?php

namespace Tomgrohl\Laravel\Encryption\Tests;

use PHPUnit_Framework_TestCase;
use Tomgrohl\Laravel\Encryption\Encrypter;

/**
 * Class EncrypterTest
 *
 * @package Tomgrohl\Laravel\Encryption\Tests
 */
class EncrypterTest extends PHPUnit_Framework_TestCase
{
    public function testEncryption()
    {
        $e = new Encrypter(str_repeat('a', 16));
        $encrypted = $e->encrypt('foo');
        $this->assertNotEquals('foo', $encrypted);
        $this->assertEquals('foo', $e->decrypt($encrypted));
    }
    public function testRawStringEncryption()
    {
        $e = new Encrypter(str_repeat('a', 16));
        $encrypted = $e->encryptString('foo');
        $this->assertNotEquals('foo', $encrypted);
        $this->assertEquals('foo', $e->decryptString($encrypted));
    }
    public function testEncryptionUsingBase64EncodedKey()
    {
        $e = new Encrypter(random_bytes(16));
        $encrypted = $e->encrypt('foo');
        $this->assertNotEquals('foo', $encrypted);
        $this->assertEquals('foo', $e->decrypt($encrypted));
    }
    public function testWithCustomCipher()
    {
        $e = new Encrypter(str_repeat('b', 32), 'AES-256-CBC');
        $encrypted = $e->encrypt('bar');
        $this->assertNotEquals('bar', $encrypted);
        $this->assertEquals('bar', $e->decrypt($encrypted));
        $e = new Encrypter(random_bytes(32), 'AES-256-CBC');
        $encrypted = $e->encrypt('foo');
        $this->assertNotEquals('foo', $encrypted);
        $this->assertEquals('foo', $e->decrypt($encrypted));
    }
    /**
     * @expectedException \RuntimeException
     * @expectedExceptionMessage The only supported ciphers are AES-128-CBC and AES-256-CBC with the correct key lengths.
     */
    public function testDoNoAllowLongerKey()
    {
        new Encrypter(str_repeat('z', 32));
    }
    /**
     * @expectedException \RuntimeException
     * @expectedExceptionMessage The only supported ciphers are AES-128-CBC and AES-256-CBC with the correct key lengths.
     */
    public function testWithBadKeyLength()
    {
        new Encrypter(str_repeat('a', 5));
    }
    /**
     * @expectedException \RuntimeException
     * @expectedExceptionMessage The only supported ciphers are AES-128-CBC and AES-256-CBC with the correct key lengths.
     */
    public function testWithBadKeyLengthAlternativeCipher()
    {
        new Encrypter(str_repeat('a', 16), 'AES-256-CFB8');
    }
    /**
     * @expectedException \RuntimeException
     * @expectedExceptionMessage The only supported ciphers are AES-128-CBC and AES-256-CBC with the correct key lengths.
     */
    public function testWithUnsupportedCipher()
    {
        new Encrypter(str_repeat('c', 16), 'AES-256-CFB8');
    }
    /**
     * @expectedException \Illuminate\Encryption\DecryptException
     * @expectedExceptionMessage The payload is invalid.
     */
    public function testExceptionThrownWhenPayloadIsInvalid()
    {
        $e = new Encrypter(str_repeat('a', 16));
        $payload = $e->encrypt('foo');
        $payload = str_shuffle($payload);
        $e->decrypt($payload);
    }
    /**
     * @expectedException \Illuminate\Encryption\DecryptException
     * @expectedExceptionMessage The MAC is invalid.
     */
    public function testExceptionThrownWithDifferentKey()
    {
        $a = new Encrypter(str_repeat('a', 16));
        $b = new Encrypter(str_repeat('b', 16));
        $b->decrypt($a->encrypt('baz'));
    }
}


src/Tomgrohl/Laravel/Encryption/Tests/EncryptionServiceProviderTest.php
<?php

namespace Tomgrohl\Laravel\Encryption\Tests;

use Illuminate\Foundation\Application;
use Tomgrohl\Laravel\Encryption\Encrypter;
use Tomgrohl\Laravel\Encryption\EncryptionServiceProvider;

/**
 * Class EncryptionServiceProvider
 *
 * @package Tomgrohl\Laravel\Encryption
 */
class EncryptionServiceProviderTest extends TestCase
{
    public function testProvider()
    {
        $application = new Application();

        $provider = new EncryptionServiceProvider($application);
        $provider->register();

        $this->assertTrue($application->bound('encrypter'));
        $this->assertTrue($application->isAlias(Encrypter::class));
    }
}


src/Tomgrohl/Laravel/Encryption/Encrypter.php
<?php

namespace Tomgrohl\Laravel\Encryption;

use Illuminate\Encryption\DecryptException;
use Illuminate\Encryption\Encrypter as BaseEncrypter;

/**
 * Class Encrypter
 *
 * A port of the Laravel 5.4 Encrypter for 4.2 with PHP 7.1 support
 *
 * @package Tomgrohl\Laravel\Encryption
 */
class Encrypter extends BaseEncrypter
{
    /**
     * Encrypter constructor.
     *
     * @param string $key
     * @param string $cipher
     */
    public function __construct($key, $cipher = 'AES-128-CBC')
    {
        parent::__construct($key);

        $this->cipher = $cipher;

        if ( ! static::supported($this->key, $this->cipher)) {
            throw new \RuntimeException('The only supported ciphers are AES-128-CBC and AES-256-CBC with the correct key lengths.');
        }
    }

    /**
     * Determine if the given key and cipher combination is valid.
     *
     * @param  string  $key
     * @param  string  $cipher
     * @return bool
     */
    public static function supported($key, $cipher)
    {
        $length = mb_strlen($key, '8bit');

        return ($cipher === 'AES-128-CBC' && $length === 16) ||
            ($cipher === 'AES-256-CBC' && $length === 32);
    }

    /**
     * @param $value
     * @param bool $serialize
     * @return string
     */
    public function encrypt($value, $serialize = true)
    {
        $iv = random_bytes(16);

        // First we will encrypt the value using OpenSSL. After this is encrypted we
        // will proceed to calculating a MAC for the encrypted value so that this
        // value can be verified later as not having been changed by the users.
        $value = \openssl_encrypt(
            $serialize ? serialize($value) : $value,
            $this->cipher, $this->key, 0, $iv
        );

        if ($value === false) {
            throw new \RuntimeException('Could not encrypt the data.');
        }

        // Once we have the encrypted value we will go ahead base64_encode the input
        // vector and create the MAC for the encrypted value so we can verify its
        // authenticity. Then, we'll JSON encode the data in a "payload" array.
        $mac = $this->hash($iv = base64_encode($iv), $value);

        $json = json_encode(compact('iv', 'value', 'mac'));

        if (! is_string($json)) {
            throw new \RuntimeException('Could not encrypt the data.');
        }

        return base64_encode($json);
    }

    /**
     * Encrypt a string without serialization.
     *
     * @param  string  $value
     * @return string
     */
    public function encryptString($value)
    {
        return $this->encrypt($value, false);
    }

    /**
     * @param $payload
     * @param bool $unserialize
     * @return mixed|string
     */
    public function decrypt($payload, $unserialize = true)
    {
        $payload = $this->getJsonPayload($payload);

        $iv = base64_decode($payload['iv']);

        // Here we will decrypt the value. If we are able to successfully decrypt it
        // we will then unserialize it and return it out to the caller. If we are
        // unable to decrypt this value we will throw out an exception message.
        try {
            $decrypted = \openssl_decrypt(
                $payload['value'], $this->cipher, $this->key, 0, $iv
            );
        } catch (\Exception $e) {
            // Pass exceptions through as DecryptExceptions so they're handled properly
            throw new DecryptException($e->getMessage());
        }

        if ($decrypted === false) {
            throw new DecryptException('Could not decrypt the data.');
        }

        return $unserialize ? unserialize($decrypted) : $decrypted;
    }

    /**
     * Decrypt the given string without unserialization.
     *
     * @param  string  $payload
     * @return string
     */
    public function decryptString($payload)
    {
        return $this->decrypt($payload, false);
    }

    /**
     * Create a MAC for the given value.
     *
     * @param  string  $iv
     * @param  mixed  $value
     * @return string
     */
    protected function hash($iv, $value)
    {
        return hash_hmac('sha256', $iv.$value, $this->key);
    }

    /**
     * @param $payload
     * @return mixed
     */
    protected function getJsonPayload($payload)
    {
        $payload = json_decode(base64_decode($payload), true);

        // If the payload is not valid JSON or does not have the proper keys set we will
        // assume it is invalid and bail out of the routine since we will not be able
        // to decrypt the given value. We'll also check the MAC for this encryption.
        if (! $this->validPayload($payload)) {
            throw new DecryptException('The payload is invalid.');
        }

        if (! $this->validMac($payload)) {
            throw new DecryptException('The MAC is invalid.');
        }

        return $payload;
    }

    /**
     * Verify that the encryption payload is valid.
     *
     * @param  mixed  $payload
     * @return bool
     */
    protected function validPayload($payload)
    {
        return is_array($payload) && isset(
                $payload['iv'], $payload['value'], $payload['mac']
            );
    }

    /**
     * Determine if the MAC for the given payload is valid.
     *
     * @param  array  $payload
     * @return bool
     */
    protected function validMac(array $payload)
    {
        $calculated = $this->calculateMac($payload, $bytes = random_bytes(16));

        return hash_equals(
            hash_hmac('sha256', $payload['mac'], $bytes, true), $calculated
        );
    }

    /**
     * Calculate the hash of the given payload.
     *
     * @param  array  $payload
     * @param  string  $bytes
     * @return string
     */
    protected function calculateMac($payload, $bytes)
    {
        return hash_hmac(
            'sha256', $this->hash($payload['iv'], $payload['value']), $bytes, true
        );
    }

    /**
     * Get the encryption key.
     *
     * @return string
     */
    public function getKey()
    {
        return $this->key;
    }
}
